infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInfrastructure.java

```diff
@@ -37,25 +42,40 @@ public OpenShiftInfrastructure(
       EventService eventService,
       OpenShiftRuntimeContextFactory runtimeContextFactory,
       OpenShiftEnvironmentProvisioner osEnvProvisioner,
-      Set<InternalEnvironmentProvisioner> internalEnvProvisioners) {
-    super(NAME, ImmutableSet.of(OpenShiftEnvironment.TYPE), eventService, internalEnvProvisioners);
+      Set<InternalEnvironmentProvisioner> internalEnvProvisioners,
+      DockerImageEnvironmentConverter dockerImageEnvConverter) {
+    super(
+        NAME,
+        ImmutableSet.of(OpenShiftEnvironment.TYPE, DockerImageEnvironment.TYPE),
+        eventService,
+        internalEnvProvisioners);
     this.runtimeContextFactory = runtimeContextFactory;
     this.osEnvProvisioner = osEnvProvisioner;
+    this.dockerImageEnvConverter = dockerImageEnvConverter;
   }
 
   @Override
   protected OpenShiftRuntimeContext internalPrepare(
       RuntimeIdentity id, InternalEnvironment environment)
       throws ValidationException, InfrastructureException {
-    if (!(environment instanceof OpenShiftEnvironment)) {
-      throw new ValidationException(
-          "Environment type '%s' is not supported. Supported environment " + "types: %s",
-          environment.getRecipe().getType(), OpenShiftEnvironment.TYPE);
-    }
-    OpenShiftEnvironment openShiftEnvironment = (OpenShiftEnvironment) environment;
+    final OpenShiftEnvironment openShiftEnvironment = asOpenShiftEnv(environment);
 
     osEnvProvisioner.provision(openShiftEnvironment, id);
 
     return runtimeContextFactory.create(openShiftEnvironment, id, this);
   }
+
+  private OpenShiftEnvironment asOpenShiftEnv(InternalEnvironment source)
+      throws ValidationException, InfrastructureException {
+    if (source instanceof OpenShiftEnvironment) {
+      return (OpenShiftEnvironment) source;
+    }
+    if (source instanceof DockerImageEnvironment) {
+      return dockerImageEnvConverter.convert((DockerImageEnvironment) source);
+    }
+    throw new ValidationException(
```

Comment:
Consider changing to InternalInfrastructureException since reaching of this code is possible in case of a developer mistake only